In [45]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sns

sns.set_theme()

# jupyter notebook full-width display
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

# no text wrapping
display(HTML("<style>.dataframe td { white-space: nowrap; }</style>"))

# pandas formatting
pd.set_option('display.float_format', '{:.3f}'.format)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 200)
pd.set_option('display.max_colwidth', 400)

In [3]:
df_SD = pd.read_pickle('df_SD.pickle')
df_FD = pd.read_pickle('df_FD.pickle')
df_LF = pd.read_pickle('df_LF.pickle')
df_LF_grouped = pd.read_pickle('df_LF_grouped.pickle')
df_Site = pd.read_pickle('df_Site.pickle')
df_TrapSupervisors = pd.read_pickle('df_TrapSupervisors.pickle')

# checking stuff

### why don't these match?
* answer: they do, but there is no matching sample to link them

In [4]:
df_LF[(df_LF.yy==2019) & (df_LF.mm==5) & (df_LF.dd==28) & (df_LF.site=="47")]

yy  mm  dd  Time     river  week site    loc period  wt_lbs  wt_kg  \
11195  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11196  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11197  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11198  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11199  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11200  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11201  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11202  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11203  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11204  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11205  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   

       lgth  freq  Flbin   DATETIME SITE1 SITE2 SITE3          id  \
11195   215     1    215 2019-05-28    47  <NA>  <NA>  2019052847   
11196   220     6    220 2019-05-28    47  <NA>  <NA>  2019052847   
11197   225    16    225 2019-05-28    47  <NA>  <NA>  2019052847   
11198   230    26    230 2019-05-28    47  <NA>  <NA>  2019052847   
11199   235    39    235 2019-05-28    47  <NA>  <NA>  2019052847   
11200   240    51    240 2019-05-28    47  <NA>  <NA>  2019052847   
11201   245    35    245 2019-05-28    47  <NA>  <NA>  2019052847   
11202   250    27    250 2019-05-28    47  <NA>  <NA>  2019052847   
11203   255    17    255 2019-05-28    47  <NA>  <NA>  2019052847   
11204   260     5    260 2019-05-28    47  <NA>  <NA>  2019052847   
11205   265     1    265 2019-05-28    47  <NA>  <NA>  2019052847   

       length_bin_id FLAG_SITE FLAG_AM_PM_PERIOD  \
11195         21.500       NaN               NaN   
11196         22.000       NaN               NaN   
11197         22.500       NaN               NaN   
11198         23.000       NaN               NaN   
11199         23.500       NaN               NaN   
11200         24.000       NaN               NaN   
11201         24.500       NaN               NaN   
11202         25.000       NaN               NaN   
11203         25.500       NaN               NaN   
11204         26.000       NaN               NaN   
11205         26.500       NaN               NaN   

      FLAG_AM_PM_PERIOD_DISCREPANCIES FLAG_NO_MATCHING_SAMPLE  
11195                             NaN                    True  
11196                             NaN                    True  
11197                             NaN                    True  
11198                             NaN                    True  
11199                             NaN                    True  
11200                             NaN                    True  
11201                             NaN                    True  
11202                             NaN                    True  
11203                             NaN                    True  
11204                             NaN                    True  
11205                             NaN                    True

In [5]:
df_SD[(df_SD.DATETIME.dt.year==2019) & (df_SD.DATETIME.dt.month==5) & (df_SD.DATETIME.dt.day==28)]

DIST   RIVER                 NAME code  GEAR SITE_NO no_nets  YEAR  MM  \
15115     2  SWMARG     Martin E Cameron  NaN    81      12       1  2019   5   
15116     2  SWMARG    Gerard MacFarlane  NaN    81      28       1  2019   5   
15135     2  SWMARG        Mary E Gillis  NaN    81      49       1  2019   5   
15136     2  SWMARG  Elizabeth MacKinnon  NaN    81      62       1  2019   5   
15137     2  SWMARG      Kevin MacKinnon  NaN    81      64       1  2019   5   
15138     2  SWMARG       Stewart Gillis  NaN    81      41       1  2019   5   
15139     2  SWMARG      Pierre Chiasson  NaN    81      26       1  2019   5   
15140     2  SWMARG      Bruce MacLellan  NaN    81      51       1  2019   5   
15141     2  SWMARG       Daniel Stewart  NaN    81      33       1  2019   5   
15142     2  SWMARG        Robert Peters  NaN    81      25       1  2019   5   
15143     2  SWMARG       Finley Stewart  NaN    81      44     NaN  2019   5   

       DD  Week  catch_lbs  catch_kg hours_fished   zone   last_name comments  \
15115  28     5    350.000   158.800            5  lower     Cameron      NaN   
15116  28     5    200.000    90.700            8  upper  MacFarlane      NaN   
15135  28     5    700.000   317.500           14  upper      Gillis      NaN   
15136  28     5    500.000   226.800           10  upper   MacKinnon      NaN   
15137  28     5    120.000    54.400            6  upper   Mackinnon      NaN   
15138  28     5     30.000    13.600            7  upper      Gillis      NaN   
15139  28     5    200.000    90.700            7  lower    Chiasson      NaN   
15140  28     5     15.000     6.800            6  upper   MacLellan      NaN   
15141  28     5    150.000    68.000            8  upper     Stewart      NaN   
15142  28     5     25.000    11.300            8  lower      Peters      NaN   
15143  28     5    300.000   136.100           14  upper     Stewart      NaN   

       bycatch_sbass  bycatch_shad bycatch_other   DATETIME  SITE1 SITE2  \
15115            NaN           NaN           NaN 2019-05-28     12  <NA>   
15116            NaN           NaN       1 perch 2019-05-28     28  <NA>   
15135            NaN           NaN           NaN 2019-05-28     49  <NA>   
15136            NaN           NaN           NaN 2019-05-28     62  <NA>   
15137            NaN           NaN           NaN 2019-05-28     64  <NA>   
15138            NaN           NaN           NaN 2019-05-28     41  <NA>   
15139            NaN           NaN           NaN 2019-05-28     26  <NA>   
15140            NaN           NaN           NaN 2019-05-28     51  <NA>   
15141            NaN           NaN           NaN 2019-05-28     33  <NA>   
15142            NaN           NaN           NaN 2019-05-28     25  <NA>   
15143            NaN           NaN           NaN 2019-05-28     44  <NA>   

                      remarks          id  total_fish_preserved  \
15115                     NaN  2019052812                31.000   
15116  bycatch_other: 1 perch  2019052828                   NaN   
15135                     NaN  2019052849                   NaN   
15136                     NaN  2019052862                36.000   
15137                     NaN  2019052864                   NaN   
15138                     NaN  2019052841                   NaN   
15139                     NaN  2019052826                   NaN   
15140                     NaN  2019052851                   NaN   
15141                     NaN  2019052833                   NaN   
15142                     NaN  2019052825                   NaN   
15143                     NaN  2019052844                   NaN   

       total_fish_measured AM_PM_PERIOD  wt_lbs FLAG_DATETIME  \
15115              195.000           AM  71.900           NaN   
15116                  NaN          NaN     NaN           NaN   
15135                  NaN          NaN     NaN           NaN   
15136              203.000           AM  74.100           NaN   
15137             

### kevin followup question: how many ambiguous matches don't have a matching sample because there are multiple sample matches

In [6]:
ambiguous = set()
for i in list(df_SD[df_SD.id>2024000000].id):
    while i > 2024000000:
        i -= 1000000000
    ambiguous.add(i)
    
ambiguous = list(ambiguous)

In [7]:
# EXACT MATCHES = NONE

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # no exact matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY) & (df_LF.site==str(SITE))])
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY) & (df_FD.SITE==SITE)])


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

In [8]:
# POTENTIAL MATCHES WITH MESSY SITE NAMES = NONE

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # partial matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY) & (df_LF.site.str.contains(str(SITE))==True)])
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY) & (df_FD.SITE.str.contains(str(SITE))==True)])


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

In [9]:
# A FEW MATCHES WHEN SITE IS NOT CONSIDERED
# maybe there the site number was written down wrong, some partial matches (excluding site)

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # very partial matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY)].head())
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY)].head())


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


yy  mm  dd  Time        river  week site    loc period  wt_lbs  wt_kg  \
3024  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3025  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3026  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3027  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3028  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   

      lgth  freq  Flbin   DATETIME SITE1 SITE2 SITE3          id  \
3024   220     3    220 1997-06-12    26  <NA>  <NA>  1997061226   
3025   225     4    225 1997-06-12    26  <NA>  <NA>  1997061226   
3026   230    19    230 1997-06-12    26  <NA>  <NA>  1997061226   
3027   235    52    235 1997-06-12    26  <NA>  <NA>  1997061226   
3028   240    70    240 1997-06-12    26  <NA>  <NA>  1997061226   

      length_bin_id FLAG_SITE FLAG_AM_PM_PERIOD  \
3024         22.000       NaN               NaN   
3025         22.500       NaN               NaN   
3026         23.000       NaN               NaN   
3027         23.500       NaN               NaN   
3028         24.000       NaN               NaN   

     FLAG_AM_PM_PERIOD_DISCREPANCIES FLAG_NO_MATCHING_SAMPLE  
3024                             NaN                     NaN  
3025                             NaN                     NaN  
3026                             NaN                     NaN  
3027                             NaN                     NaN  
3028                             NaN                     NaN

YEAR  MM  DD  WEEK SITE PERIOD CONDITION  FISH_NO  FL_WET  FL_FROZEN  \
index                                                                         
15211  1997   6  12     7   26     AM    Frozen        1    <NA>        230   
15212  1997   6  12     7   26     AM    Frozen       10    <NA>        237   
15213  1997   6  12     7   26     AM    Frozen       11    <NA>        246   
15214  1997   6  12     7   26     AM    Frozen       12    <NA>        251   
15215  1997   6  12     7   26     AM    Frozen       13    <NA>        228   

       FL_STD  WEIGHT SPECIES SEX  MATURITY  GONAD_WEIGHT Ager_1  AGE_1  \
index                                                                     
15211 238.000 171.000       A   M         4           NaN    NaN      3   
15212 245.000 182.000       A   M         4           NaN    NaN      3   
15213 254.000 231.000       A   F         4        31.600    NaN      3   
15214 259.000 211.000       A   F         4        34.200    NaN      3   
15215 236.000 172.000       A   F         4        28.500    NaN      3   

       FSP_1 Comments_1 Ager_2  AGE_2  FSP_2 Comments_2 Ager_3  AGE_3  FSP_3  \
index                                                                          
15211      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15212      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15213      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15214      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15215      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   

      Comments_3 Envelop.Comments   DATETIME AGE_notes_1 FSP_notes_1  \
index                                                                  
15211        NaN              NaN 1997-06-12           3           3   
15212        NaN              NaN 1997-06-12           3           3   
15213        NaN              NaN 1997-06-12           3           3   
15214        NaN              NaN 1997-06-12           3           3   
15215        NaN              NaN 1997-06-12           3           3   

      AGE_notes_2 FSP_notes_2 AGE_notes_3 FSP_notes_3 SITE_notes  SITE1 SITE2  \
index                                                                           
15211         NaN         NaN         NaN         NaN         26     26  <NA>   
15212         NaN         NaN         NaN         NaN         26     26  <NA>   
15213         NaN         NaN         NaN         NaN         26     26  <NA>   
15214         NaN         NaN         NaN         NaN         26     26  <NA>   
15215         NaN         NaN         NaN         NaN         26     26  <NA>   

      SITE3  fish_length                                         remarks  \
index                                                                      
15211  <NA>          230  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15212  <NA>          237  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15213  <NA>          246  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15214  <NA>          251  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15215  <NA>          228  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   

               id FLAG_SITE FLAG_AM_PM_PERIOD FLAG_SEX FLAG_MATURITY  \
index                                                                  
15211  1997061226       NaN               NaN      NaN           NaN   
15212  1997061226       NaN               NaN      NaN           NaN   
15213  1997061226       NaN               NaN      NaN           NaN   
15214  1997061226       NaN               NaN      NaN           NaN   
15215  1997061226       NaN               NaN      NaN           NaN   

      FLAG_FSP_1 FLAG_FL_STD FLAG_FL_WET_FROZEN FLAG_WEIGHT_OUTLIER  \
index                                                                 
15211        NaN         NaN                NaN                 NaN   
15212        NaN         NaN              


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

YEAR  MM  DD  WEEK SITE PERIOD CONDITION  FISH_NO  FL_WET  FL_FROZEN  \
index                                                                         
4408   1988   5  23  <NA>   12    NaN     Fresh        1     253       <NA>   
4409   1988   5  23  <NA>   12    NaN     Fresh       10     250       <NA>   
4410   1988   5  23  <NA>   12    NaN     Fresh       11     253       <NA>   
4411   1988   5  23  <NA>   12    NaN     Fresh       12     248       <NA>   
4412   1988   5  23  <NA>   12    NaN     Fresh       13     245       <NA>   

       FL_STD  WEIGHT SPECIES SEX  MATURITY  GONAD_WEIGHT Ager_1  AGE_1  \
index                                                                     
4408  253.000 208.000       A   M         4           NaN    NaN      4   
4409  250.000 217.000       A   M         4           NaN    NaN      4   
4410  253.000 198.000       A   M         4           NaN    NaN      4   
4411  248.000 193.000       A   M         4           NaN    NaN      4   
4412  245.000 199.000       A   M         4           NaN    NaN      3   

       FSP_1 Comments_1 Ager_2  AGE_2  FSP_2 Comments_2 Ager_3  AGE_3  FSP_3  \
index                                                                          
4408       4        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4409       3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4410       3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4411       3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4412       3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   

      Comments_3 Envelop.Comments   DATETIME AGE_notes_1 FSP_notes_1  \
index                                                                  
4408         NaN              NaN 1988-05-23           4           4   
4409         NaN              NaN 1988-05-23           4           3   
4410         NaN              NaN 1988-05-23           4           3   
4411         NaN              NaN 1988-05-23           4           3   
4412         NaN              NaN 1988-05-23           3           3   

      AGE_notes_2 FSP_notes_2 AGE_notes_3 FSP_notes_3 SITE_notes  SITE1 SITE2  \
index                                                                           
4408          NaN         NaN         NaN         NaN         12     12  <NA>   
4409          NaN         NaN         NaN         NaN         12     12  <NA>   
4410          NaN         NaN         NaN         NaN         12     12  <NA>   
4411          NaN         NaN         NaN         NaN         12     12  <NA>   
4412          NaN         NaN         NaN         NaN         12     12  <NA>   

      SITE3  fish_length                                         remarks  \
index                                                                      
4408   <NA>          253  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4409   <NA>          250  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 3   
4410   <NA>          253  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 3   
4411   <NA>          248  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 3   
4412   <NA>          245  SITE_notes: 12; AGE_notes_1: 3; FSP_notes_1: 3   

               id FLAG_SITE FLAG_AM_PM_PERIOD FLAG_SEX FLAG_MATURITY  \
index                                                                  
4408   1988052312       NaN               NaN      NaN           NaN   
4409   1988052312       NaN               NaN      NaN           NaN   
4410   1988052312       NaN               NaN      NaN           NaN   
4411   1988052312       NaN               NaN      NaN           NaN   
4412   1988052312       NaN               NaN      NaN           NaN   

      FLAG_FSP_1 FLAG_FL_STD FLAG_FL_WET_FROZEN FLAG_WEIGHT_OUTLIER  \
index                                                                 
4408         NaN         NaN                NaN                 NaN   
4409         NaN         NaN              


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

### how do we flag and leave the potential to merge ghost/ambiguous samples?

In [10]:
id = 1990061312
date = id // 100
date

19900613

In [11]:
from datetime import datetime
datetime.strptime(str(id//100), "%Y%m%d").date()

datetime.date(1990, 6, 13)

In [12]:
id = 4990061312
date = id // 100
while date > 20240000:
    date -= 10000000
datetime.strptime(str(date), "%Y%m%d").date()

datetime.date(1990, 6, 13)

In [13]:
sum(df_LF.id.isna()), sum(df_FD.id.isna())

(0, 0)

In [14]:
id - 100*(id//100)

12

# When making ghost samples, will any of our ambiguous id match between fish details and length frequencies?

In [15]:
ambiguous_FD = set(df_FD[df_FD.id>2024000000].id)
ambiguous_LF = set(df_LF[df_LF.id>2024000000].id)
ambiguous_SD = set(df_SD[df_SD.id>2024000000].id)
# yes, we need to further disambiguate: maybe add 20/40 to months, this number would never occur naturally

In [16]:
# let's make a summary of ambiguous dates

df_A_FD = pd.DataFrame(ambiguous_FD, columns=['id_FD'])
df_A_FD['date'] = df_A_FD['id_FD'] // 100
while df_A_FD.date.max() > 20240000:
    df_A_FD.loc[df_A_FD.date>20240000, 'date'] -= 10000000
    
df_A_LF = pd.DataFrame(ambiguous_LF, columns=['id_LF'])
df_A_LF['date'] = df_A_LF['id_LF'] // 100
while df_A_LF.date.max() > 20240000:
    df_A_LF.loc[df_A_LF.date>20240000, 'date'] -= 10000000
    
df_A_SD = pd.DataFrame(ambiguous_SD, columns=['id_SD'])
df_A_SD['date'] = df_A_SD['id_SD'] // 100
while df_A_SD.date.max() > 20240000:
    df_A_SD.loc[df_A_SD.date>20240000, 'date'] -= 10000000

# group and split into columns
df_A_FD = df_A_FD.groupby('date')['id_FD'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_FD.columns = ['id_FD' for _ in range(df_A_FD.shape[1])]

df_A_LF = df_A_LF.groupby('date')['id_LF'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_LF.columns = ['id_LF' for _ in range(df_A_LF.shape[1])]

df_A_SD = df_A_SD.groupby('date')['id_SD'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_SD.columns = ['id_SD' for _ in range(df_A_SD.shape[1])]

# create a dataframe summarising ambiguous data
df_AMBIGUOUS = pd.merge(
    df_A_FD,
    pd.merge(
        df_A_SD,
        df_A_LF,
        on='date', 
        how='outer'
    ),
    on='date', 
    how='outer'
).sort_index().astype('Int64')

df_AMBIGUOUS.loc[df_AMBIGUOUS['id_FD'].any(axis=1), 'FLAG_FD_AMBIGUOUS'] = True
df_AMBIGUOUS.loc[df_AMBIGUOUS['id_LF'].any(axis=1), 'FLAG_LF_AMBIGUOUS'] = True
df_AMBIGUOUS.loc[df_AMBIGUOUS['id_SD'].any(axis=1), 'FLAG_SD_AMBIGUOUS'] = True

df_AMBIGUOUS['FLAG_AMBIGUITY_OVERLAP'] = df_AMBIGUOUS.FLAG_FD_AMBIGUOUS.notnull()*1 + df_AMBIGUOUS.FLAG_LF_AMBIGUOUS.notnull()*1 + df_AMBIGUOUS.FLAG_SD_AMBIGUOUS.notnull()*1

In [17]:
# these are our problems if we are making ghost samples

df_AMBIGUOUS.loc[df_AMBIGUOUS.FLAG_AMBIGUITY_OVERLAP > 1].dropna(axis=1).T

date                      19980501    20000607    20000609    20010516  \
id_FD                   2998050105  3000060726  3000060926  3001051605   
id_LF                   3998050105  4000060726  3000060926  3001051605   
id_LF                   2998050105  3000060726  4000060926  4001051605   
FLAG_FD_AMBIGUOUS             True        True        True        True   
FLAG_LF_AMBIGUOUS             True        True        True        True   
FLAG_AMBIGUITY_OVERLAP           2           2           2           2   

date                      20090520    20090522    20100511    20100519  \
id_FD                   3009052026  3009052226  3010051126  3010051926   
id_LF                   3009052026  3009052226  3010051126  3010051926   
id_LF                   4009052026  4009052226  4010051126  4010051926   
FLAG_FD_AMBIGUOUS             True        True        True        True   
FLAG_LF_AMBIGUOUS             True        True        True        True   
FLAG_AMBIGUITY_OVERLAP           2           2           2           2   

date                      20100527  
id_FD                   3010052726  
id_LF                   3010052726  
id_LF                   4010052726  
FLAG_FD_AMBIGUOUS             True  
FLAG_LF_AMBIGUOUS             True  
FLAG_AMBIGUITY_OVERLAP           2

# Potential Sorting Issues? LF? FD?
* Yes, can't disambiguate without making guesses. Left flagged, and will be rejected on import.

In [18]:
for i in ambiguous_FD:
    display(df_FD[df_FD.id==i][['YEAR', 'MM', 'DD', 'SITE', 'PERIOD', 'FISH_NO', 'id']].head(1))

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
5355   1989   5  14   12     AM        1  2989051412

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
5395   1989   5  14   12     AM       48  3989051412

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
11142  1993   5  29   52     AM        1  2993052952

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
18346  2000   6   7   26     AM        1  3000060726

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
18382  2000   6   7   26     PM        1  4000060726

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
24451  2010   5  11   26     AM        1  3010051126

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
15403  1998   5   1    5     PM        1  2998050105

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
24471  2010   5  11   26     AM       28  4010051126

YEAR  MM  DD      SITE PERIOD  FISH_NO          id
index                                                    
29431  2014   5  30  47 or 62     PM        1  3014053047

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
29461  2014   5  30   47     AM        1  4014053047

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
25102  2010   5  29   25     PM        1  3010052925

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
11030  1993   5  27   37     AM        1  2993052737

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
23846  2009   5  22   26     AM        1  3009052226

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
23870  2009   5  22   26     AM       31  4009052226

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
11513  1993   6   9   33     AM        1  2993060933

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
5309   1989   5  13   35     PM        1  3989051335

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
5263   1989   5  13   35     AM        1  2989051335

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
23753  2009   5  20   26     AM        1  3009052026

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
18501  2001   5  16    5     PM        1  3001051605

YEAR  MM  DD  SITE PERIOD  FISH_NO          id
index                                                
18546  2001   5  16  5,26     AM        1  4001051605

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
18427  2000   6   9   26     AM        1  3000060926

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
24634  2010   5  19   26     AM        1  3010051926

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
24671  2010   5  19   26     PM        1  4010051926

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
25450  2010   6  11   41     AM        1  3010061141

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
25227  2010   5   5   41     AM       33  3010050541

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
24982  2010   5  27   26     AM        1  3010052726

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
25011  2010   5  27   26     PM        1  4010052726

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
23782  2009   5  20   26     PM        1  4009052026

YEAR  MM  DD SITE PERIOD  FISH_NO          id
index                                               
18461  2000   6   9   26     PM        1  4000060926

In [19]:
# CONFIRM
# some have am / pm that could be used to match with LF manually if that info is there as well???
for i in ambiguous_FD:
    print('\n', i)
    display(df_FD[df_FD.id==i][['YEAR', 'MM', 'DD', 'SITE', 'PERIOD', 'FISH_NO']].head())


 2989051412


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
5355   1989   5  14   12     AM        1
5356   1989   5  14   12     AM       10
5357   1989   5  14   12     AM       11
5358   1989   5  14   12     AM       12
5359   1989   5  14   12     AM       13


 3989051412


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
5395   1989   5  14   12     AM       48
5398   1989   5  14   12     AM       50
5403   1989   5  14   12     PM        1
5404   1989   5  14   12     PM       10
5405   1989   5  14   12     PM       11


 2993052952


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
11142  1993   5  29   52     AM        1
11143  1993   5  29   52     AM       10
11144  1993   5  29   52     AM       11
11145  1993   5  29   52     AM       11
11146  1993   5  29   52     AM       12


 3000060726


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
18346  2000   6   7   26     AM        1
18347  2000   6   7   26     AM       10
18348  2000   6   7   26     AM       11
18349  2000   6   7   26     AM       12
18350  2000   6   7   26     AM       13


 4000060726


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
18382  2000   6   7   26     PM        1
18383  2000   6   7   26     PM       10
18384  2000   6   7   26     PM       11
18385  2000   6   7   26     PM       12
18386  2000   6   7   26     PM       13


 3010051126


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
24451  2010   5  11   26     AM        1
24452  2010   5  11   26     AM       10
24453  2010   5  11   26     AM       11
24454  2010   5  11   26     AM       12
24455  2010   5  11   26     AM       13


 2998050105


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
15403  1998   5   1    5     PM        1
15404  1998   5   1    5     PM        1
15405  1998   5   1    5     PM       10
15406  1998   5   1    5     PM       10
15407  1998   5   1    5     PM       11


 4010051126


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
24471  2010   5  11   26     AM       28
24480  2010   5  11   26     PM        1
24481  2010   5  11   26     PM       10
24482  2010   5  11   26     PM       11
24483  2010   5  11   26     PM       12


 3014053047


YEAR  MM  DD      SITE PERIOD  FISH_NO
index                                        
29431  2014   5  30  47 or 62     PM        1
29432  2014   5  30  47 or 62     PM       10
29433  2014   5  30  47 or 62     PM       11
29434  2014   5  30  47 or 62     PM       12
29435  2014   5  30  47 or 62     PM       13


 4014053047


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
29461  2014   5  30   47     AM        1
29462  2014   5  30   47     AM       10
29463  2014   5  30   47     AM       11
29464  2014   5  30   47     AM       12
29465  2014   5  30   47     AM       13


 3010052925


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
25102  2010   5  29   25     PM        1
25103  2010   5  29   25     PM       10
25104  2010   5  29   25     PM       11
25105  2010   5  29   25     PM       12
25106  2010   5  29   25     PM       13


 2993052737


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
11030  1993   5  27   37     AM        1
11031  1993   5  27   37     AM       10
11032  1993   5  27   37     AM       11
11033  1993   5  27   37     AM       12
11034  1993   5  27   37     AM       13


 3009052226


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
23846  2009   5  22   26     AM        1
23847  2009   5  22   26     AM       10
23848  2009   5  22   26     AM       11
23849  2009   5  22   26     AM       12
23850  2009   5  22   26     AM       13


 4009052226


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
23870  2009   5  22   26     AM       31
23877  2009   5  22   26     PM        1
23878  2009   5  22   26     PM       10
23879  2009   5  22   26     PM       11
23880  2009   5  22   26     PM       12


 2993060933


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
11513  1993   6   9   33     AM        1
11514  1993   6   9   33     AM       10
11515  1993   6   9   33     AM       11
11516  1993   6   9   33     AM       12
11517  1993   6   9   33     AM       13


 3989051335


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
5309   1989   5  13   35     PM        1
5310   1989   5  13   35     PM       10
5311   1989   5  13   35     PM       11
5312   1989   5  13   35     PM       12
5313   1989   5  13   35     PM       13


 2989051335


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
5263   1989   5  13   35     AM        1
5264   1989   5  13   35     AM       10
5265   1989   5  13   35     AM       11
5266   1989   5  13   35     AM       12
5267   1989   5  13   35     AM       13


 3009052026


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
23753  2009   5  20   26     AM        1
23754  2009   5  20   26     AM       10
23755  2009   5  20   26     AM       11
23756  2009   5  20   26     AM       12
23757  2009   5  20   26     AM       13


 3001051605


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
18501  2001   5  16    5     PM        1
18502  2001   5  16    5     PM       10
18503  2001   5  16    5     PM       11
18504  2001   5  16    5     PM       12
18505  2001   5  16    5     PM       13


 4001051605


YEAR  MM  DD  SITE PERIOD  FISH_NO
index                                    
18546  2001   5  16  5,26     AM        1
18547  2001   5  16  5,26     AM       10
18548  2001   5  16  5,26     AM       11
18549  2001   5  16  5,26     AM       12
18550  2001   5  16  5,26     AM       13


 3000060926


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
18427  2000   6   9   26     AM        1
18428  2000   6   9   26     AM       10
18429  2000   6   9   26     AM       11
18430  2000   6   9   26     AM       12
18431  2000   6   9   26     AM       13


 3010051926


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
24634  2010   5  19   26     AM        1
24635  2010   5  19   26     AM       10
24636  2010   5  19   26     AM       11
24637  2010   5  19   26     AM       12
24638  2010   5  19   26     AM       13


 4010051926


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
24671  2010   5  19   26     PM        1
24672  2010   5  19   26     PM       10
24673  2010   5  19   26     PM       11
24674  2010   5  19   26     PM       12
24675  2010   5  19   26     PM       13


 3010061141


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
25450  2010   6  11   41     AM        1
25451  2010   6  11   41     AM       10
25452  2010   6  11   41     AM       11
25453  2010   6  11   41     AM       12
25454  2010   6  11   41     AM       13


 3010050541


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
25227  2010   5   5   41     AM       33
25228  2010   5   5   41     PM        1
25229  2010   5   5   41     PM       10
25230  2010   5   5   41     PM       11
25231  2010   5   5   41     PM       12


 3010052726


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
24982  2010   5  27   26     AM        1
24983  2010   5  27   26     AM       10
24984  2010   5  27   26     AM       11
24985  2010   5  27   26     AM       12
24986  2010   5  27   26     AM       13


 4010052726


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
25011  2010   5  27   26     PM        1
25012  2010   5  27   26     PM       10
25013  2010   5  27   26     PM       11
25014  2010   5  27   26     PM       12
25015  2010   5  27   26     PM       13


 4009052026


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
23782  2009   5  20   26     PM        1
23783  2009   5  20   26     PM       10
23784  2009   5  20   26     PM       11
23785  2009   5  20   26     PM       12
23786  2009   5  20   26     PM       13


 4000060926


YEAR  MM  DD SITE PERIOD  FISH_NO
index                                   
18461  2000   6   9   26     PM        1
18462  2000   6   9   26     PM       10
18463  2000   6   9   26     PM       11
18464  2000   6   9   26     PM       12
18465  2000   6   9   26     PM       13

In [20]:
# check ambiguous AM PM for LF and FD and look for extra matches (NOTE, none match samples/logbook)
# [x for x in df_AMBIGUOUS[['id_LF']].values.T.ravel().tolist() if not pd.isna(x)]

period_check = pd.merge(
    df_FD.loc[df_FD.id.isin([x for x in df_AMBIGUOUS[['id_FD']].values.T.ravel().tolist() if not pd.isna(x)])].groupby(['id', 'PERIOD']).count().reset_index()[['id', 'PERIOD']],
    df_LF.loc[df_LF.id.isin([x for x in df_AMBIGUOUS[['id_LF']].values.T.ravel().tolist() if not pd.isna(x)])].groupby(['id', 'period']).count().reset_index()[['id', 'period']],
    on='id',
    how='outer'
)

period_check['id'] -= 1000000000
period_check.loc[period_check.id > 2024000000, 'id'] -= 1000000000

period_check = period_check.sort_values(['id', 'PERIOD']).reset_index(drop=True)

In [21]:
id_period_ambiguity = list(period_check.dropna()[period_check.dropna().duplicated('id', keep=False)].id.unique())

check_ids = [x+1000000000 for x in id_period_ambiguity] + [x+2000000000 for x in id_period_ambiguity]

In [22]:
df_LF[df_LF.id.isin(check_ids)].groupby(['id', 'period']).first().sort_values('DATETIME')

yy  mm  dd     Time        river  week  site    loc  \
id         period                                                          
3000060726 AM      2000   6   7      NaN  SW MARGAREE     6    26  LOWER   
4000060726 PM      2000   6   7      NaN  SW MARGAREE     6    26  LOWER   
3000060926 PM      2000   6   9      NaN  SW MARGAREE     6    26  LOWER   
4000060926 AM      2000   6   9      NaN  SW MARGAREE     6    26  LOWER   
3001051605 AM      2001   5  16      NaN  SW MARGAREE     3  5,26  LOWER   
4001051605 PM      2001   5  16      NaN  SW MARGAREE     3     5  LOWER   
3009052026 AM      2009   5  20 1255.000  SW MARGAREE     4    26  LOWER   
4009052026 PM      2009   5  20 1405.000  SW MARGAREE     4    26  LOWER   
3009052226 AM      2009   5  22 1310.000  SW MARGAREE     4    26  LOWER   
4009052226 PM      2009   5  22 1455.000  SW MARGAREE     4    26  LOWER   
3010051126 AM      2010   5  11 1450.000  SW MARGAREE     3    26  LOWER   
4010051126 PM      2010   5  11 1540.000  SW MARGAREE     3    26  LOWER   
3010051926 AM      2010   5  19 1345.000  SW MARGAREE     4    26  LOWER   
4010051926 PM      2010   5  19 1630.000  SW MARGAREE     4    26  LOWER   
3010052726 AM      2010   5  27  840.000  SW MARGAREE     5    26  LOWER   
4010052726 PM      2010   5  27 1500.000  SW MARGAREE     5    26  LOWER   

                   wt_lbs  wt_kg  lgth  freq  Flbin            DATETIME SITE1  \
id         period                                                               
3000060726 AM     158.000 71.700   215     1    215 2000-06-07 00:00:00    26   
4000060726 PM     158.000 71.700   215     3    215 2000-06-07 00:00:00    26   
3000060926 PM     141.000 64.000   215     3    215 2000-06-09 00:00:00    26   
4000060926 AM     146.000 66.200   215     1    215 2000-06-09 00:00:00    26   
3001051605 AM     140.000 63.500   220     1    220 2001-05-16 00:00:00     5   
4001051605 PM     154.000 69.900   225     2    225 2001-05-16 00:00:00     5   
3009052026 AM     139.000 63.000   220     1    220 2009-05-20 12:55:00    26   
4009052026 PM      96.000 43.500   225     1    225 2009-05-20 14:05:00    26   
3009052226 AM     128.000 58.100   225     2    225 2009-05-22 13:10:00    26   
4009052226 PM     130.000 59.000   225     3    225 2009-05-22 14:55:00    26   
3010051126 AM      43.000 19.500   230     3    230 2010-05-11 14:50:00    26   
4010051126 PM      36.000 16.300   235     1    235 2010-05-11 15:40:00    26   
3010051926 AM     129.000 58.500   225     2    225 2010-05-19 13:45:00    26   
4010051926 PM     100.000 45.400   220     1    220 2010-05-19 16:30:00    26   
3010052726 AM     117.000 53.100   225     1    225 2010-05-27 08:40:00    26   
4010052726 PM     121.000 54.900   230    19    230 2010-05-27 15:00:00    26   

                  SITE2 SITE3  length_bin_id FLAG_SITE FLAG_AM_PM_PERIOD  \
id         period                                                          
3000060726 AM      None  None         21.500      None              None   
4000060726 PM      None  None         21.500      None              None   
3000060926 PM      None  None         21.500      None              None   
4000060926 AM      None  None         21.500      None              None   
3001051605 AM        26  None         22.000      True              None   
4001051605 PM      None  None         22.500      None              None   
3009052026 AM      None  None         22.000      None              None   
4009052026 PM      None  None         22.500      None              None   
3009052226 AM      None  None         22.500      None              None   
4009052226 PM      None  None         22.500      None              None   
3010051126 AM      None  None         23.000      None              None   
4010051126 PM      None  None         23.500      None              None   
3010051926 AM      None  None         22.500      None              None   
4010051926 PM      None  None         22.000      Non

In [23]:
df_FD[df_FD.id.isin(check_ids)].groupby(['id', 'PERIOD']).first().sort_values('DATETIME')

YEAR  MM  DD  WEEK  SITE CONDITION  FISH_NO  FL_WET  \
id         PERIOD                                                        
3000060726 AM      2000   6   7     6    26    Frozen        1    <NA>   
           PM      2000   6   7     6    26     Fresh       37     242   
4000060726 PM      2000   6   7     6    26     Fresh        1     252   
3000060926 AM      2000   6   9     6    26     Fresh        1     239   
           PM      2000   6   9     6    26    Frozen       35    <NA>   
4000060926 PM      2000   6   9     6    26    Frozen        1    <NA>   
3001051605 AM      2001   5  16     3  5,26    Frozen       46    <NA>   
           PM      2001   5  16     3     5    Frozen        1    <NA>   
4001051605 AM      2001   5  16     3  5,26    Frozen        1    <NA>   
3009052026 AM      2009   5  20     4    26     Fresh        1     245   
           PM      2009   5  20     4    26    Frozen       30    <NA>   
4009052026 PM      2009   5  20     4    26    Frozen        1    <NA>   
3009052226 AM      2009   5  22     4    26     Fresh        1     232   
4009052226 AM      2009   5  22     4    26     Fresh       31     247   
           PM      2009   5  22     4    26    Frozen        1    <NA>   
4010051126 PM      2010   5  11     3    26    Frozen        1    <NA>   
3010051126 AM      2010   5  11     3    26    Frozen        1    <NA>   
4010051126 AM      2010   5  11     3    26    Frozen       28    <NA>   
3010051926 AM      2010   5  19     4    26    Frozen        1    <NA>   
4010051926 PM      2010   5  19     4    26    Frozen        1    <NA>   
3010052726 AM      2010   5  27     5    26     Fresh        1     235   
4010052726 PM      2010   5  27     5    26    Frozen        1    <NA>   

                   FL_FROZEN  FL_STD  WEIGHT SPECIES SEX  MATURITY  \
id         PERIOD                                                    
3000060726 AM            250 258.000 220.000       B   M         4   
           PM           <NA> 242.000 198.000       B   M         4   
4000060726 PM           <NA> 252.000 226.000       B   F         4   
3000060926 AM           <NA> 239.000 168.000       B   M         4   
           PM            228 236.000 154.800       A   M         4   
4000060926 PM            266 274.000 284.500       A   F         4   
3001051605 AM            217 225.000 128.400       A   M         4   
           PM            276 285.000 312.600       A   F         4   
4001051605 AM            271 279.000 288.000       A   M         4   
3009052026 AM            269 245.000 202.500       A   M         4   
           PM            265 273.000 257.800       A   F         4   
4009052026 PM            243 251.000 200.800       A   M         4   
3009052226 AM           <NA> 232.000 156.000       A   M         4   
4009052226 AM           <NA> 247.000 225.400       A   F         4   
           PM            226 234.000 160.200       A   M         4   
4010051126 PM            242 250.000 172.100       A   M         4   
3010051126 AM            251 259.000 215.500       A   M         4   
4010051126 AM            260 268.000 234.000       A   M         4   
3010051926 AM            273 281.000 271.100       A   F         4   
4010051926 PM            235 243.000 184.100       A   F         4   
3010052726 AM           <NA> 235.000 163.100       A   M         4   
4010052726 PM            239 247.000 191.700       A   M         4   

                   GONAD_WEIGHT Ager_1  AGE_1  FSP_1 Comments_1 Ager_2  AGE_2  \
id         PERIOD                                                               
3000060726 AM            28.700   None      3      9       None   None   <NA>   
           PM            52.500   None      3      9       None   None   <NA>   
4000060726 PM            28.800   None      3      9       None   None   <NA>   
3000060926 AM            34.000   None      3      0       None   None   <NA>   
           PM            35.200   None   <NA>      9       None   None   <N

4 ambiguous samples may be linked:

    3010051926	AM
    3010051926	AM
    
    4010051926	PM
    4010051926	PM
    
    3010052726	AM
    3010052726	AM
    
    4010052726	PM
    4010052726	PM
    
Note: these are matched together correctly.

* Therefore, 

        if id.isin([3010051926, 4010051926, 3010052726, 4010052726]):
            don't add anything to months to disambiguate 
            # these id will match up correctly and link to the same ghost sample

* Other id from this set do not match unambiguously using period. There are some inconsistencies.
* In fact, period may not be entirely useful because of the inconsistencies with within ambiguous sets of fish details.

# More Sorting Issue Calcs and Checks

In [24]:
# loop through duplicated ids and look at these by inspection - make sure they all look doubled (except 2018061364)
# the only one that looks like a typo is 2018061364 (fixed at "CONFIRM: fix typo at 10851" in noteboook 10.1)

# from notebook 10.1
duplicated_ids_LF = [1998050105, 2000060726, 2000060926, 2000061538, 2001051605, 2003052326,
                     2009052026, 2009052226, 2010051126, 2010051926, 2010052726, 2018061364]

temp = df_LF.copy()
temp.loc[temp.id>2024000000, 'id'] -= 1000000000
temp.loc[temp.id>2024000000, 'id'] -= 1000000000

for duplicate_id in duplicated_ids_LF:
    idx = temp[temp.id == duplicate_id].index.values[0]
    
    before = 0
    after = temp[temp.id == duplicate_id].shape[0]-1
    
    print('\n', temp[temp.id == duplicate_id]['id'].values[0])
    display(temp.loc[idx-before:idx+after][['id', 'DATETIME', 'site', 'period', 'lgth', 'wt_lbs']])



 1998050105


id   DATETIME site period  lgth  wt_lbs
3102  1998050105 1998-05-01    5     PM   235  10.000
3103  1998050105 1998-05-01    5     PM   240  10.000
3104  1998050105 1998-05-01    5     PM   245  10.000
3105  1998050105 1998-05-01    5     PM   250  10.000
3106  1998050105 1998-05-01    5     PM   255  10.000
3107  1998050105 1998-05-01    5     PM   260  10.000
3108  1998050105 1998-05-01    5     PM   265  10.000
3109  1998050105 1998-05-01    5     PM   270  10.000
3110  1998050105 1998-05-01    5     PM   275  10.000
3111  1998050105 1998-05-01    5     PM   280  10.000
3112  1998050105 1998-05-01    5     AM   230 132.000
3113  1998050105 1998-05-01    5     AM   235 132.000
3114  1998050105 1998-05-01    5     AM   240 132.000
3115  1998050105 1998-05-01    5     AM   245 132.000
3116  1998050105 1998-05-01    5     AM   250 132.000
3117  1998050105 1998-05-01    5     AM   255 132.000
3118  1998050105 1998-05-01    5     AM   260 132.000
3119  1998050105 1998-05-01    5     AM   265 132.000
3120  1998050105 1998-05-01    5     AM   270 132.000
3121  1998050105 1998-05-01    5     AM   275 132.000
3122  1998050105 1998-05-01    5     AM   280 132.000


 2000060726


id   DATETIME site period  lgth  wt_lbs
4156  2000060726 2000-06-07   26     AM   215 158.000
4157  2000060726 2000-06-07   26     AM   220 158.000
4158  2000060726 2000-06-07   26     AM   225 158.000
4159  2000060726 2000-06-07   26     AM   230 158.000
4160  2000060726 2000-06-07   26     AM   235 158.000
4161  2000060726 2000-06-07   26     AM   240 158.000
4162  2000060726 2000-06-07   26     AM   245 158.000
4163  2000060726 2000-06-07   26     AM   250 158.000
4164  2000060726 2000-06-07   26     AM   255 158.000
4165  2000060726 2000-06-07   26     AM   260 158.000
4166  2000060726 2000-06-07   26     AM   265 158.000
4167  2000060726 2000-06-07   26     AM   270 158.000
4168  2000060726 2000-06-07   26     AM   275 158.000
4169  2000060726 2000-06-07   26     AM   280 158.000
4170  2000060726 2000-06-07   26     AM   285 158.000
4171  2000060726 2000-06-07   26     PM   215 158.000
4172  2000060726 2000-06-07   26     PM   220 158.000
4173  2000060726 2000-06-07   26     PM   225 158.000
4174  2000060726 2000-06-07   26     PM   230 158.000
4175  2000060726 2000-06-07   26     PM   235 158.000
4176  2000060726 2000-06-07   26     PM   240 158.000
4177  2000060726 2000-06-07   26     PM   245 158.000
4178  2000060726 2000-06-07   26     PM   250 158.000
4179  2000060726 2000-06-07   26     PM   255 158.000
4180  2000060726 2000-06-07   26     PM   260 158.000
4181  2000060726 2000-06-07   26     PM   265 158.000
4182  2000060726 2000-06-07   26     PM   270 158.000
4183  2000060726 2000-06-07   26     PM   275 158.000
4184  2000060726 2000-06-07   26     PM   280 158.000
4185  2000060726 2000-06-07   26     PM   285 158.000


 2000060926


id   DATETIME site period  lgth  wt_lbs
4186  2000060926 2000-06-09   26     PM   215 141.000
4187  2000060926 2000-06-09   26     PM   220 141.000
4188  2000060926 2000-06-09   26     PM   225 141.000
4189  2000060926 2000-06-09   26     PM   230 141.000
4190  2000060926 2000-06-09   26     PM   235 141.000
4191  2000060926 2000-06-09   26     PM   240 141.000
4192  2000060926 2000-06-09   26     PM   245 141.000
4193  2000060926 2000-06-09   26     PM   250 141.000
4194  2000060926 2000-06-09   26     PM   255 141.000
4195  2000060926 2000-06-09   26     PM   260 141.000
4196  2000060926 2000-06-09   26     PM   265 141.000
4197  2000060926 2000-06-09   26     PM   270 141.000
4198  2000060926 2000-06-09   26     PM   275 141.000
4199  2000060926 2000-06-09   26     PM   280 141.000
4200  2000060926 2000-06-09   26     AM   215 146.000
4201  2000060926 2000-06-09   26     AM   220 146.000
4202  2000060926 2000-06-09   26     AM   225 146.000
4203  2000060926 2000-06-09   26     AM   230 146.000
4204  2000060926 2000-06-09   26     AM   235 146.000
4205  2000060926 2000-06-09   26     AM   240 146.000
4206  2000060926 2000-06-09   26     AM   245 146.000
4207  2000060926 2000-06-09   26     AM   250 146.000
4208  2000060926 2000-06-09   26     AM   255 146.000
4209  2000060926 2000-06-09   26     AM   260 146.000
4210  2000060926 2000-06-09   26     AM   265 146.000
4211  2000060926 2000-06-09   26     AM   270 146.000
4212  2000060926 2000-06-09   26     AM   275 146.000


 2000061538


id   DATETIME site period  lgth  wt_lbs
4239  2000061538 2000-06-15   38     PM   205 139.000
4240  2000061538 2000-06-15   38     PM   210 139.000
4241  2000061538 2000-06-15   38     PM   215 139.000
4242  2000061538 2000-06-15   38     PM   220 139.000
4243  2000061538 2000-06-15   38     PM   225 139.000
4244  2000061538 2000-06-15   38     PM   230 139.000
4245  2000061538 2000-06-15   38     PM   235 139.000
4246  2000061538 2000-06-15   38     PM   240 139.000
4247  2000061538 2000-06-15   38     PM   245 139.000
4248  2000061538 2000-06-15   38     PM   250 139.000
4249  2000061538 2000-06-15   38     PM   255 139.000
4250  2000061538 2000-06-15   38     PM   260 139.000
4251  2000061538 2000-06-15   38     PM   265 139.000
4252  2000061538 2000-06-15   38     PM   270 139.000
4253  2000061538 2000-06-15   38     PM   275 139.000
4254  2000061538 2000-06-15   38     AM   220  57.000
4255  2000061538 2000-06-15   38     AM   225  57.000
4256  2000061538 2000-06-15   38     AM   230  57.000
4257  2000061538 2000-06-15   38     AM   235  57.000
4258  2000061538 2000-06-15   38     AM   240  57.000
4259  2000061538 2000-06-15   38     AM   245  57.000
4260  2000061538 2000-06-15   38     AM   250  57.000
4261  2000061538 2000-06-15   38     AM   255  57.000
4262  2000061538 2000-06-15   38     AM   260  57.000
4263  2000061538 2000-06-15   38     AM   265  57.000
4264  2000061538 2000-06-15   38     AM   270  57.000
4265  2000061538 2000-06-15   38     AM   275  57.000


 2001051605


id   DATETIME  site period  lgth  wt_lbs
4266  2001051605 2001-05-16  5,26     AM   220 140.000
4267  2001051605 2001-05-16  5,26     AM   225 140.000
4268  2001051605 2001-05-16  5,26     AM   230 140.000
4269  2001051605 2001-05-16  5,26     AM   235 140.000
4270  2001051605 2001-05-16  5,26     AM   240 140.000
4271  2001051605 2001-05-16  5,26     AM   245 140.000
4272  2001051605 2001-05-16  5,26     AM   250 140.000
4273  2001051605 2001-05-16  5,26     AM   255 140.000
4274  2001051605 2001-05-16  5,26     AM   260 140.000
4275  2001051605 2001-05-16  5,26     AM   265 140.000
4276  2001051605 2001-05-16  5,26     AM   270 140.000
4277  2001051605 2001-05-16  5,26     AM   275 140.000
4278  2001051605 2001-05-16  5,26     AM   280 140.000
4279  2001051605 2001-05-16  5,26     AM   285 140.000
4280  2001051605 2001-05-16  5,26     AM   295 140.000
4281  2001051605 2001-05-16     5     PM   225 154.000
4282  2001051605 2001-05-16     5     PM   230 154.000
4283  2001051605 2001-05-16     5     PM   235 154.000
4284  2001051605 2001-05-16     5     PM   240 154.000
4285  2001051605 2001-05-16     5     PM   245 154.000
4286  2001051605 2001-05-16     5     PM   250 154.000
4287  2001051605 2001-05-16     5     PM   255 154.000
4288  2001051605 2001-05-16     5     PM   260 154.000
4289  2001051605 2001-05-16     5     PM   265 154.000
4290  2001051605 2001-05-16     5     PM   270 154.000
4291  2001051605 2001-05-16     5     PM   275 154.000
4292  2001051605 2001-05-16     5     PM   280 154.000
4293  2001051605 2001-05-16     5     PM   285 154.000
4294  2001051605 2001-05-16     5     PM   290 154.000
4295  2001051605 2001-05-16     5     PM   295 154.000


 2003052326


id   DATETIME site period  lgth  wt_lbs
5184  2003052326 2003-05-23   26     AM   230  30.000
5185  2003052326 2003-05-23   26     AM   235  30.000
5186  2003052326 2003-05-23   26     AM   240  30.000
5187  2003052326 2003-05-23   26     AM   245  30.000
5188  2003052326 2003-05-23   26     AM   250  30.000
5189  2003052326 2003-05-23   26     AM   255  30.000
5190  2003052326 2003-05-23   26     AM   260  30.000
5191  2003052326 2003-05-23   26     AM   265  30.000
5192  2003052326 2003-05-23   26     AM   270  30.000
5193  2003052326 2003-05-23   26     AM   275  30.000
5194  2003052326 2003-05-23   26     PM   225  47.000
5195  2003052326 2003-05-23   26     PM   230  47.000
5196  2003052326 2003-05-23   26     PM   235  47.000
5197  2003052326 2003-05-23   26     PM   240  47.000
5198  2003052326 2003-05-23   26     PM   245  47.000
5199  2003052326 2003-05-23   26     PM   250  47.000
5200  2003052326 2003-05-23   26     PM   255  47.000
5201  2003052326 2003-05-23   26     PM   260  47.000
5202  2003052326 2003-05-23   26     PM   265  47.000
5203  2003052326 2003-05-23   26     PM   270  47.000
5204  2003052326 2003-05-23   26     PM   275  47.000
5205  2003052326 2003-05-23   26     PM   280  47.000


 2009052026


id            DATETIME site period  lgth  wt_lbs
6332  2009052026 2009-05-20 12:55:00   26     AM   220 139.000
6333  2009052026 2009-05-20 12:55:00   26     AM   225 139.000
6334  2009052026 2009-05-20 12:55:00   26     AM   230 139.000
6335  2009052026 2009-05-20 12:55:00   26     AM   235 139.000
6336  2009052026 2009-05-20 12:55:00   26     AM   240 139.000
6337  2009052026 2009-05-20 12:55:00   26     AM   245 139.000
6338  2009052026 2009-05-20 12:55:00   26     AM   250 139.000
6339  2009052026 2009-05-20 12:55:00   26     AM   255 139.000
6340  2009052026 2009-05-20 12:55:00   26     AM   260 139.000
6341  2009052026 2009-05-20 12:55:00   26     AM   265 139.000
6342  2009052026 2009-05-20 12:55:00   26     AM   270 139.000
6343  2009052026 2009-05-20 14:05:00   26     PM   225  96.000
6344  2009052026 2009-05-20 14:05:00   26     PM   230  96.000
6345  2009052026 2009-05-20 14:05:00   26     PM   235  96.000
6346  2009052026 2009-05-20 14:05:00   26     PM   240  96.000
6347  2009052026 2009-05-20 14:05:00   26     PM   245  96.000
6348  2009052026 2009-05-20 14:05:00   26     PM   250  96.000
6349  2009052026 2009-05-20 14:05:00   26     PM   255  96.000
6350  2009052026 2009-05-20 14:05:00   26     PM   260  96.000
6351  2009052026 2009-05-20 14:05:00   26     PM   265  96.000
6352  2009052026 2009-05-20 14:05:00   26     PM   270  96.000
6353  2009052026 2009-05-20 14:05:00   26     PM   275  96.000
6354  2009052026 2009-05-20 14:05:00   26     PM   280  96.000


 2009052226


id            DATETIME site period  lgth  wt_lbs
6368  2009052226 2009-05-22 13:10:00   26     AM   225 128.000
6369  2009052226 2009-05-22 13:10:00   26     AM   230 128.000
6370  2009052226 2009-05-22 13:10:00   26     AM   235 128.000
6371  2009052226 2009-05-22 13:10:00   26     AM   240 128.000
6372  2009052226 2009-05-22 13:10:00   26     AM   245 128.000
6373  2009052226 2009-05-22 13:10:00   26     AM   250 128.000
6374  2009052226 2009-05-22 13:10:00   26     AM   255 128.000
6375  2009052226 2009-05-22 13:10:00   26     AM   260 128.000
6376  2009052226 2009-05-22 13:10:00   26     AM   265 128.000
6377  2009052226 2009-05-22 13:10:00   26     AM   270 128.000
6378  2009052226 2009-05-22 13:10:00   26     AM   275 128.000
6379  2009052226 2009-05-22 14:55:00   26     PM   225 130.000
6380  2009052226 2009-05-22 14:55:00   26     PM   230 130.000
6381  2009052226 2009-05-22 14:55:00   26     PM   235 130.000
6382  2009052226 2009-05-22 14:55:00   26     PM   240 130.000
6383  2009052226 2009-05-22 14:55:00   26     PM   245 130.000
6384  2009052226 2009-05-22 14:55:00   26     PM   250 130.000
6385  2009052226 2009-05-22 14:55:00   26     PM   255 130.000
6386  2009052226 2009-05-22 14:55:00   26     PM   260 130.000
6387  2009052226 2009-05-22 14:55:00   26     PM   265 130.000
6388  2009052226 2009-05-22 14:55:00   26     PM   270 130.000


 2010051126


id            DATETIME site period  lgth  wt_lbs
6648  2010051126 2010-05-11 14:50:00   26     AM   230  43.000
6649  2010051126 2010-05-11 14:50:00   26     AM   235  43.000
6650  2010051126 2010-05-11 14:50:00   26     AM   240  43.000
6651  2010051126 2010-05-11 14:50:00   26     AM   245  43.000
6652  2010051126 2010-05-11 14:50:00   26     AM   250  43.000
6653  2010051126 2010-05-11 14:50:00   26     AM   255  43.000
6654  2010051126 2010-05-11 14:50:00   26     AM   260  43.000
6655  2010051126 2010-05-11 14:50:00   26     AM   265  43.000
6656  2010051126 2010-05-11 14:50:00   26     AM   270  43.000
6657  2010051126 2010-05-11 14:50:00   26     AM   275  43.000
6658  2010051126 2010-05-11 14:50:00   26     AM   280  43.000
6659  2010051126 2010-05-11 15:40:00   26     PM   235  36.000
6660  2010051126 2010-05-11 15:40:00   26     PM   240  36.000
6661  2010051126 2010-05-11 15:40:00   26     PM   245  36.000
6662  2010051126 2010-05-11 15:40:00   26     PM   250  36.000
6663  2010051126 2010-05-11 15:40:00   26     PM   255  36.000
6664  2010051126 2010-05-11 15:40:00   26     PM   260  36.000
6665  2010051126 2010-05-11 15:40:00   26     PM   265  36.000
6666  2010051126 2010-05-11 15:40:00   26     PM   270  36.000
6667  2010051126 2010-05-11 15:40:00   26     PM   275  36.000
6668  2010051126 2010-05-11 15:40:00   26     PM   280  36.000
6669  2010051126 2010-05-11 15:40:00   26     PM   285  36.000


 2010051926


id            DATETIME site period  lgth  wt_lbs
6719  2010051926 2010-05-19 13:45:00   26     AM   225 129.000
6720  2010051926 2010-05-19 13:45:00   26     AM   230 129.000
6721  2010051926 2010-05-19 13:45:00   26     AM   235 129.000
6722  2010051926 2010-05-19 13:45:00   26     AM   240 129.000
6723  2010051926 2010-05-19 13:45:00   26     AM   245 129.000
6724  2010051926 2010-05-19 13:45:00   26     AM   250 129.000
6725  2010051926 2010-05-19 13:45:00   26     AM   255 129.000
6726  2010051926 2010-05-19 13:45:00   26     AM   260 129.000
6727  2010051926 2010-05-19 13:45:00   26     AM   265 129.000
6728  2010051926 2010-05-19 13:45:00   26     AM   270 129.000
6729  2010051926 2010-05-19 13:45:00   26     AM   275 129.000
6730  2010051926 2010-05-19 13:45:00   26     AM   280 129.000
6731  2010051926 2010-05-19 13:45:00   26     AM   285 129.000
6732  2010051926 2010-05-19 16:30:00   26     PM   220 100.000
6733  2010051926 2010-05-19 16:30:00   26     PM   225 100.000
6734  2010051926 2010-05-19 16:30:00   26     PM   230 100.000
6735  2010051926 2010-05-19 16:30:00   26     PM   235 100.000
6736  2010051926 2010-05-19 16:30:00   26     PM   240 100.000
6737  2010051926 2010-05-19 16:30:00   26     PM   245 100.000
6738  2010051926 2010-05-19 16:30:00   26     PM   250 100.000
6739  2010051926 2010-05-19 16:30:00   26     PM   255 100.000
6740  2010051926 2010-05-19 16:30:00   26     PM   260 100.000
6741  2010051926 2010-05-19 16:30:00   26     PM   265 100.000
6742  2010051926 2010-05-19 16:30:00   26     PM   270 100.000
6743  2010051926 2010-05-19 16:30:00   26     PM   275 100.000
6744  2010051926 2010-05-19 16:30:00   26     PM   280 100.000


 2010052726


id            DATETIME site period  lgth  wt_lbs
6846  2010052726 2010-05-27 08:40:00   26     AM   225 117.000
6847  2010052726 2010-05-27 08:40:00   26     AM   230 117.000
6848  2010052726 2010-05-27 08:40:00   26     AM   235 117.000
6849  2010052726 2010-05-27 08:40:00   26     AM   240 117.000
6850  2010052726 2010-05-27 08:40:00   26     AM   245 117.000
6851  2010052726 2010-05-27 08:40:00   26     AM   250 117.000
6852  2010052726 2010-05-27 08:40:00   26     AM   255 117.000
6853  2010052726 2010-05-27 08:40:00   26     AM   260 117.000
6854  2010052726 2010-05-27 08:40:00   26     AM   265 117.000
6855  2010052726 2010-05-27 08:40:00   26     AM   270 117.000
6856  2010052726 2010-05-27 08:40:00   26     AM   275 117.000
6857  2010052726 2010-05-27 15:00:00   26     PM   230 121.000
6858  2010052726 2010-05-27 15:00:00   26     PM   235 121.000
6859  2010052726 2010-05-27 15:00:00   26     PM   240 121.000
6860  2010052726 2010-05-27 15:00:00   26     PM   245 121.000
6861  2010052726 2010-05-27 15:00:00   26     PM   250 121.000
6862  2010052726 2010-05-27 15:00:00   26     PM   255 121.000
6863  2010052726 2010-05-27 15:00:00   26     PM   260 121.000
6864  2010052726 2010-05-27 15:00:00   26     PM   265 121.000
6865  2010052726 2010-05-27 15:00:00   26     PM   270 121.000
6866  2010052726 2010-05-27 15:00:00   26     PM   275 121.000
6867  2010052726 2010-05-27 15:00:00   26     PM   280 121.000


 2018061364


id   DATETIME site period  lgth  wt_lbs
10841  2018061364 2018-06-13   64     PM   215  92.000
10842  2018061364 2018-06-13   64     PM   220  92.000
10843  2018061364 2018-06-13   64     PM   225  92.000
10844  2018061364 2018-06-13   64     PM   230  92.000
10845  2018061364 2018-06-13   64     PM   235  92.000
10846  2018061364 2018-06-13   64     PM   240  92.000
10847  2018061364 2018-06-13   64     PM   245  92.000
10848  2018061364 2018-06-13   64     PM   250  92.000
10849  2018061364 2018-06-13   64     PM   255  92.000
10850  2018061364 2018-06-13   64     PM   260  92.000
10851  2018061364 2018-06-13   64     PM   265  92.000
10852  2018061364 2018-06-13   64     PM   270  92.000

In [25]:
# it looks like 4018061364 is our only potential error (fixed earlier, now 2018061364)
df_LF.groupby('id').count().sort_values('yy')['yy'].head()

id
2007060235    1
1991051212    3
1994050848    4
2005050501    6
1991051312    6
Name: yy, dtype: int64

In [26]:
# loop through small counts above and look at them
# these look fine, unless there were multiple combined typos
smallgroups = list(df_LF.groupby('id').count().sort_values('yy')['yy'].head().index)

for group in smallgroups:
    idx = df_LF[df_LF.id == group].index.values[0]
    before = 2
    after = 8
    
    print('\n', df_LF[df_LF.id == group]['id'].values[0])
    display(df_LF.loc[idx-before:idx+after][['DATETIME', 'site', 'period', 'lgth', 'wt_lbs']])



 2007060235


DATETIME site period  lgth  wt_lbs
5960 2007-06-01   47     PM   285 133.000
5961 2007-06-01   47     PM   295 133.000
5962 2007-06-02   35     PM   225 122.000
5963 2007-06-02    2     PM   230 122.000
5964 2007-06-02    2     PM   235 122.000
5965 2007-06-02    2     PM   240 122.000
5966 2007-06-02    2     PM   245 122.000
5967 2007-06-02    2     PM   250 122.000
5968 2007-06-02    2     PM   255 122.000
5969 2007-06-02    2     PM   260 122.000
5970 2007-06-02    2     PM   265 122.000


 1991051212


DATETIME site period  lgth  wt_lbs
946 1990-06-08 00:00:00   64     AM   285     NaN
947 1990-06-08 00:00:00   64     AM   293     NaN
948 1991-05-12 03:00:00   12     AM   265   1.000
949 1991-05-12 03:00:00   12     AM   270   1.000
950 1991-05-12 03:00:00   12     AM   275   1.000
951 1991-05-13 03:00:00   12     AM   265  11.000
952 1991-05-13 03:00:00   12     AM   270  11.000
953 1991-05-13 03:00:00   12     AM   275  11.000
954 1991-05-13 03:00:00   12     AM   280  11.000
955 1991-05-13 03:00:00   12     AM   285  11.000
956 1991-05-13 03:00:00   12     AM   290  11.000


 1994050848


DATETIME site period  lgth  wt_lbs
10341 1994-06-01   41     AM   280 137.000
10342 1994-06-01   41     AM   300 137.000
10343 1994-05-08   48     AM   250   3.000
10344 1994-05-08   48     AM   255   3.000
10345 1994-05-08   48     AM   260   3.000
10346 1994-05-08   48     AM   265   3.000
10347 1994-06-02   49     PM   235 118.000
10348 1994-06-02   49     PM   240 118.000
10349 1994-06-02   49     PM   245 118.000
10350 1994-06-02   49     PM   250 118.000
10351 1994-06-02   49     PM   255 118.000


 2005050501


DATETIME site period  lgth  wt_lbs
5570 2004-06-16    5     PM   260  41.000
5571 2004-06-16    5     PM   265  41.000
5572 2005-05-05    1     PM   240   9.000
5573 2005-05-05    1     PM   250   9.000
5574 2005-05-05    1     PM   255   9.000
5575 2005-05-05    1     PM   260   9.000
5576 2005-05-05    1     PM   265   9.000
5577 2005-05-05    1     PM   270   9.000
5578 2005-05-07   26     PM   240  12.000
5579 2005-05-07   26     PM   245  12.000
5580 2005-05-07   26     PM   250  12.000


 1991051312


DATETIME site period  lgth  wt_lbs
949 1991-05-12 03:00:00   12     AM   270   1.000
950 1991-05-12 03:00:00   12     AM   275   1.000
951 1991-05-13 03:00:00   12     AM   265  11.000
952 1991-05-13 03:00:00   12     AM   270  11.000
953 1991-05-13 03:00:00   12     AM   275  11.000
954 1991-05-13 03:00:00   12     AM   280  11.000
955 1991-05-13 03:00:00   12     AM   285  11.000
956 1991-05-13 03:00:00   12     AM   290  11.000
957 1991-05-16 03:00:00   12     PM   255  11.000
958 1991-05-16 03:00:00   12     PM   260  11.000
959 1991-05-16 03:00:00   12     PM   265  11.000

# Trap Supervisors Linked to Sites

In [27]:
# looks correct

RELEVANT_SUPERVISORS = df_SD.groupby(['NAME', 'SITE1']).count().reset_index()[['SITE1', 'NAME']].sort_values('SITE1').reset_index(drop=True).rename({'SITE1': 'site'}, axis=1)

for site in RELEVANT_SUPERVISORS.site.unique():
    print('\nSite', site, '--', ', '.join(list(RELEVANT_SUPERVISORS.query('site == @site').NAME)))


Site 1 -- Marilyn Gillis, Charles McDaniel

Site 2 -- Richard Gillis

Site 4 -- Charles McDaniel

Site 5 -- Anthony Cameron

Site 6 -- Irene MacIsaac

Site 7 -- Eleanor MacLellan

Site 8 -- Colin Gillis

Site 9 -- John A Chisholm

Site 11 -- Darlene Cameron, Josh Fraser, Joshua Fraser

Site 12 -- Martin E Cameron

Site 15 -- Pat Chisholm, John A Chisholm

Site 17 -- Pat Chisholm, John A Chisholm

Site 18 -- Eleanor McDaniel, Brian Doyle

Site 19 -- John A Chisholm

Site 20 -- John Coady

Site 21 -- Jim Coady

Site 23 -- John A Coady

Site 24 -- James A Hirtle

Site 25 -- Robert Peters

Site 26 -- Gerard V Chiasson, Pierre Chiasson

Site 27 -- Leo Chiasson

Site 28 -- Gerard MacFarlane

Site 29 -- Donald MacLeod

Site 30 -- Robert MacLeod, Donald MacLeod

Site 31 -- Robert MacLeod

Site 32 -- Robert MacLeod, Catherine MacLeod

Site 33 -- Daniel Stewart, Michael J Stewart

Site 34 -- Joan Ingram, Fred Ingram, Ronald J Stewart

Site 35 -- John MacLellan, Jack MacLellan

Site 37 -- Floren

In [28]:
# looks correct

for supervisor in RELEVANT_SUPERVISORS.sort_values('NAME').NAME.unique():
    print(supervisor, '--', ', '.join(str(x) for x in list(RELEVANT_SUPERVISORS.query('NAME == @supervisor').site.unique())))

Alexander Gillis -- 56
Alexander MacDonald -- 39
Allan B Gillis -- 46
Anthony Cameron -- 5
Brian Doyle -- 18
Brian MacFarlane -- 66
Bruce MacLellan -- 51
Bruce McLellan -- 51
Catherine MacFarlane -- 52
Catherine MacLeod -- 32
Charles McDaniel -- 1, 4
Chris MacLean -- 68
Colin Gillis -- 8, 90
Daniel Stewart -- 33
Darlene Cameron -- 11
David MacKinnon -- 39
Donald D Gillis -- 43
Donald J Gillis -- 37
Donald M Campbell -- 45
Donald MacEachern -- 63
Donald MacLeod -- 29, 30
Donelda M Gillis -- 47
Eleanor MacLellan -- 7
Eleanor McDaniel -- 18, 91
Elizabeth MacKinnon -- 62
Elizabeth/Vincent MacKinnon -- 62
Finley MacDonald -- 62
Finley Stewart -- 44
Florence Gillis -- 37
Fred Ingram -- 34
Gerard MacFarlane -- 28
Gerard V Chiasson -- 26
Harold MacFarlane -- 67
Hugh J Gillis -- 38
Hughie MacDonnell -- 61
Irene MacIsaac -- 6
Jack MacLellan -- 35
James A Hirtle -- 24
James MacFarlane -- 55, 65
Jim Coady -- 21
Jim MacFarlane -- 62
Joan Ingram -- 34
John A Chisholm -- 9, 15, 17, 19
John A Coady --

# check more stuff

In [33]:
oldid = 2019050749
df_SD.query('id == @oldid', engine='python')

DIST   RIVER           NAME code  GEAR SITE_NO no_nets  YEAR  MM  DD  \
14890     2  SWMARG  Mary E Gillis  NaN    81      49       1  2019   5   7   

       Week  catch_lbs  catch_kg hours_fished   zone last_name comments  \
14890     2    150.000    68.000            7  upper    Gillis      NaN   

       bycatch_sbass  bycatch_shad bycatch_other   DATETIME  SITE1 SITE2  \
14890            NaN           NaN           NaN 2019-05-07     49  <NA>   

      remarks          id  total_fish_preserved  total_fish_measured  \
14890     NaN  2019050749                   NaN                  NaN   

      AM_PM_PERIOD  wt_lbs FLAG_DATETIME FLAG_HOURS_FISHED FLAG_SITE  \
14890          NaN     NaN           NaN               NaN       NaN   

      FLAG_AM_PM_PERIOD_DISCREPANCIES  
14890                             NaN

In [35]:
dtypes = {
    'code': 'object',
    'no_nets': 'object',
    'YEAR': 'Int32',
    'MM': 'Int32',
    'DD': 'Int32',
    'Week': 'Int32',
#     'hours_fished': 'Int32'  # some entries say 'maximum '
}
df_SD_orig = pd.read_csv(r'.\provided data\Margaree Gaspereau logbooks_Master.csv', dtype=dtypes)

In [37]:
df_SD_orig.query('YEAR == 2019 and MM == 5 and DD == 7 and SITE_NO == "49" ', engine='python')

DIST   RIVER           NAME code  GEAR SITE_NO no_nets  YEAR  MM  DD  \
14890     2  SWMARG  Mary E Gillis  NaN    81      49       1  2019   5   7   

       Week  catch_lbs  catch_kg hours_fished   zone last_name comments  \
14890     2    150.000    68.000            7  upper    Gillis      NaN   

       bycatch_sbass  bycatch_shad bycatch_other  
14890            NaN           NaN           NaN

# Comparing Comments with Provided Data

In [60]:
dtypes_FD = dict(df_FD.dtypes)
dtypes_LF = dict(df_LF.dtypes)
dtypes_SD = dict(df_SD.dtypes)

del dtypes_FD['DATETIME']
del dtypes_LF['DATETIME']
del dtypes_SD['DATETIME']

df2FD = pd.read_csv(r'.\provided data\revised with comments\gaspereau_fish_details_AD.csv', dtype=dtypes_FD)
df2LF = pd.read_csv(r'.\provided data\revised with comments\gaspereau_length_frequencies_AD.csv', dtype=dtypes_LF)
df2SD = pd.read_csv(r'.\provided data\revised with comments\gaspereau_sample_data_AD.csv', dtype=dtypes_SD)

df2FD['DATETIME'] = pd.to_datetime(df2FD['DATETIME'])
df2LF['DATETIME'] = pd.to_datetime(df2LF['DATETIME'])
df2SD['DATETIME'] = pd.to_datetime(df2SD['DATETIME'])

In [243]:
# id are different because of recent disambiguation for ghost sampels:
# drop id columns when checking for changes
(
    (df_FD.id.astype(str).str[1:] == df2FD.id.astype(str).str[1:]).all(),
    (df_LF.id.astype(str).str[1:] == df2LF.id.astype(str).str[1:]).all(), 
    (df_SD.id.astype(str).str[1:] == df2SD.id.astype(str).str[1:]).all()
)

(True, True, True)

### Fish Details

In [268]:
# cast everything to a string to compare differences
df_FD.index = df_FD.index.rename('')

# drop new columns and id column
drop_FD = [x for x in list(df_FD.columns) if x not in list(df2FD.columns)]

FD_diff = (
    df_FD.drop(drop_FD+['id'], axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
    != df2FD.drop('id', axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
)

FD_diff_columns = list()
for i, row in pd.DataFrame(FD_diff.any()).reset_index().rename({'index':'field', 0:'differences'}, axis=1).iterrows():
    if row['differences']:
        FD_diff_columns.append(row['field'])

In [269]:
FD_differences = pd.concat([
    df2FD[FD_diff][FD_diff.any(axis=1)][FD_diff_columns],
    df_FD.drop(drop_FD, axis=1)[FD_diff][FD_diff.any(axis=1)][FD_diff_columns]
], axis=1)

In [271]:
FD_differences.astype(str).replace({'TRUE':'True','<NA>':'','nan':''})

FL_WET FL_FROZEN    FL_STD MATURITY SITE2 FLAG_AM_PM_PERIOD FL_WET  \
14053                                         2                            
14054                                         2                            
14055                                         2                            
14056                                         2                            
14057                                         2                            
14058                                         2                            
14059                                         2                            
14060                                         2                            
14061                                         2                            
14062                                         2                            
14063                                         2                            
14064                                         2                            
14065                                         2                            
14066                                         2                            
14067                                         2                            
14068                                         2                            
14069                                         2                            
14070                                         2                            
14071                                         2                            
14072                                         2                            
14073                                         2                            
14074                                         2                            
14075                                         2                            
14076                                         2                            
14077                                         2                            
14078                                         2                            
14079                                         2                            
14080                                         2                            
14081                                         2                            
14082                                         2                            
14083                                         2                            
14084                                         2                            
14085                                         2                            
14086                                         2                            
14087                                         2                            
14088                                         2                            
15925                                                       AM/PM          
15926                                                       AM/PM          
15927                                                       AM/PM          
15928                                                       AM/PM          
15929                                                       AM/PM          
15930                                                       AM/PM          
15931                                                       AM/PM          
15932                                                       AM/PM          
15933                                                       AM/PM          
15934                                                       AM/PM          
15935                                                       AM/PM          
15936                                                       AM/PM          
15937                                                       AM/PM          
15938                                                       AM/PM          
15939                                                       AM/PM          
15940                                                 

### Length Frequencies

In [273]:
LF_diff = (
    df_LF.drop('id', axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
    != df2LF.drop('id', axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
)

LF_diff_columns = list()
for i, row in pd.DataFrame(LF_diff.any()).reset_index().rename({'index':'field', 0:'differences'}, axis=1).iterrows():
    if row['differences']:
        LF_diff_columns.append(row['field'])

In [275]:
LF_differences = pd.concat([
    df2LF[LF_diff][LF_diff.any(axis=1)][LF_diff_columns],
    df_LF[LF_diff][LF_diff.any(axis=1)][LF_diff_columns]
], axis=1)

In [276]:
# Perfect. Already got this one.
LF_differences.astype(str).replace({'TRUE':'True','<NA>':'','nan':''})

lgth  Flbin length_bin_id   lgth  Flbin length_bin_id
10851  235.0  235.0          23.5  265.0  265.0          26.5

### Samples

In [279]:
[x for x in list(df2SD.columns) if x not in list(df_SD.columns)]

['period']

In [281]:
[x for x in list(df_SD.columns) if x not in list(df2SD.columns)]

['AM_PM_PERIOD']

In [299]:
# slightly better merge with newer version (AM_PM_PERIOD)
# => drop these columns
pd.concat([
    df_SD['AM_PM_PERIOD'].fillna(0), 
    df2SD['period'].fillna(0)], 
    axis=1
)[(df_SD['AM_PM_PERIOD'].fillna(0) != df2SD['period'].fillna(0))]

AM_PM_PERIOD period
3077            AM      0
3090            PM      0
3102            PM      0
3110            PM      0
3119            AM      0
3121            PM      0
3122            AM      0
3124            PM      0
3135            AM      0
3144            PM      0
3157            AM      0
3171            AM      0
3175            PM      0
3204            PM      0
3210            PM      0
3234            PM      0
3237            AM      0
3246            AM      0
3262            PM      0
3267            AM      0
3270            PM      0
3305            AM      0
3306            AM      0
3327            PM      0
3330            AM      0
3350            PM      0
3356            AM      0
3361            AM      0
3368            AM      0
3374            AM      0
3375            AM      0
3379            AM      0
3381            PM      0
3382            PM      0
3392            AM      0
3402            AM      0
3419            PM      0
3437            AM      0
3754            PM      0
5130            AM     PM
5257            AM      0
5347            AM      0
5485            AM      0
5651            AM      0
5653            PM      0
5671            AM      0
5823            AM      0
6132            PM      0
6551            AM      0
7288            AM      0
7909            PM      0
8200            AM     PM
8330            AM      0
8422            PM      0
8940            AM      0
9381            PM      0
10211           AM     PM
10370           PM      0
11069           AM      0
11789           PM      0
11942           AM      0
12005           AM      0
13373           AM     PM
13452           PM      0
14740           PM      0

In [300]:
# cast everything to a string to compare differences
df_SD.index = df_SD.index.rename('')

# drop new columns and id column
drop_SD = [x for x in list(df_SD.columns) if x not in list(df2SD.columns)]

SD_diff = (
    df_SD.drop(['id', 'AM_PM_PERIOD'], axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
    != df2SD.drop(['id', 'period'], axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
)

SD_diff_columns = list()
for i, row in pd.DataFrame(SD_diff.any()).reset_index().rename({'index':'field', 0:'differences'}, axis=1).iterrows():
    if row['differences']:
        SD_diff_columns.append(row['field'])

In [301]:
SD_differences = pd.concat([
    df2SD[SD_diff][SD_diff.any(axis=1)][SD_diff_columns],
    df_SD[SD_diff][SD_diff.any(axis=1)][SD_diff_columns]
], axis=1)

In [302]:
# nothing new here
SD_differences.astype(str).replace({'TRUE':'True','<NA>':'','nan':''})

Empty DataFrame
Columns: []
Index: []

In [307]:
SD_diff.any()

DIST                               False
RIVER                              False
NAME                               False
code                               False
GEAR                               False
SITE_NO                            False
no_nets                            False
YEAR                               False
MM                                 False
DD                                 False
Week                               False
catch_lbs                          False
catch_kg                           False
hours_fished                       False
zone                               False
last_name                          False
comments                           False
bycatch_sbass                      False
bycatch_shad                       False
bycatch_other                      False
DATETIME                           False
SITE1                              False
SITE2                              False
remarks                            False
total_fish_prese